src/crypto/PSASpake2p.h

```diff
@@ -0,0 +1,156 @@
+     * @param my_identity_len   The verifier identity length.
+     * @param peer_identity     The peer identity. May be NULL if identities are not established.
+     * @param peer_identity_len The peer identity length.
+     * @param w0in              The input w0 (a parameter baked into the device or computed with ComputeW0).
+     * @param w0in_len          The input w0 length.
+     * @param Lin               The input L (a parameter baked into the device or computed with ComputeL).
+     * @param Lin_len           The input L length.
+     *
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    CHIP_ERROR BeginVerifier(const uint8_t * my_identity, size_t my_identity_len, const uint8_t * peer_identity,
+                             size_t peer_identity_len, const uint8_t * w0in, size_t w0in_len, const uint8_t * Lin, size_t Lin_len);
+
+    /**
+     * @brief Start the Spake2+ process as a prover (i.e. a commissioner).
+     *
+     * @param my_identity       The prover identity. May be NULL if identities are not established.
+     * @param my_identity_len   The prover identity length.
+     * @param peer_identity     The peer identity. May be NULL if identities are not established.
+     * @param peer_identity_len The peer identity length.
+     * @param w0in              The input w0 (an output from the PBKDF).
+     * @param w0in_len          The input w0 length.
+     * @param w1in              The input w1 (an output from the PBKDF).
+     * @param w1in_len          The input w1 length.
+     *
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    CHIP_ERROR BeginProver(const uint8_t * my_identity, size_t my_identity_len, const uint8_t * peer_identity,
+                           size_t peer_identity_len, const uint8_t * w0in, size_t w0in_len, const uint8_t * w1in, size_t w1in_len);
+
+    /**
+     * @brief Compute the first round of the protocol.
+     *
+     * @param pab      X value from commissioner.
+     * @param pab_len  X length.
+     * @param out     The output first round Spake2+ contribution.
+     * @param out_len The output first round Spake2+ contribution length.
+     *
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    CHIP_ERROR ComputeRoundOne(const uint8_t * pab, size_t pab_len, uint8_t * out, size_t * out_len);
+
+    /**
+     * @brief Compute the second round of the protocol.
+     *
+     * @param in      The peer first round Spake2+ contribution.
+     * @param in_len  The peer first round Spake2+ contribution length.
+     * @param out     The output second round Spake2+ contribution.
+     * @param out_len The output second round Spake2+ contribution length.
+     *
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    CHIP_ERROR ComputeRoundTwo(const uint8_t * in, size_t in_len, uint8_t * out, size_t * out_len);
```

Comment:
Is `out_len` in-out parameter? Looking at the implementation it seems to me the `out_len` should have a valid value determining the max allowed size of `out` buffer and on return it will be filled with the number of bytes returned. 
If this is correct maybe it would not harm to add such information to the docs (or split max size and output length into 2 parameters)

Reply:
This interface mimics existing Spake2p class which indeed treats out_len as in-out parameter. But I can extend the docs, sure.